openmmtools/integrators.py

```diff
@@ -1185,25 +1095,565 @@ def substep_function(step_string):
+                                                                    constraint_tolerance=constraint_tolerance,
+                                                                    measure_shadow_work=measure_shadow_work,
+                                                                    measure_heat=measure_heat,
+                                                                    )
+
+    def addUpdateAlchemicalParametersStep(self):
+        """
+        Update Context parameters according to provided functions.
+        """
+        for context_parameter in self._alchemical_functions:
+            if context_parameter in self._system_parameters:
+                self.addComputeGlobal(context_parameter, self._alchemical_functions[context_parameter])
+
+    def addAlchemicalPerturbationStep(self):
+        """
+        Add alchemical perturbation step, accumulating protocol work.
+        """
+        # Store initial potential energy
+        self.addComputeGlobal("Eold", "energy")
+
+        # Use fractional state
+        if self._direction == 'forward':
+            self.addComputeGlobal('lambda', '(step+1)/nsteps')
+        elif self._direction == 'reverse':
+            self.addComputeGlobal('lambda', '(nsteps - step - 1)/nsteps')
+
+        # Update all slaved alchemical parameters
+        self.addUpdateAlchemicalParametersStep()
+
+        # Accumulate protocol work
+        self.addComputeGlobal("Enew", "energy")
+        self.addComputeGlobal("protocol_work", "protocol_work + (Enew-Eold)/kT")
+
+    def sanity_check(self, splitting, allowed_characters="H{}RVO0123456789"):
+        super(AlchemicalLangevinSplittingIntegrator, self).sanity_check(splitting, allowed_characters=allowed_characters)
+
+    def substep_function(self, step_string, measure_shadow_work, measure_heat, n_R, force_group_nV, mts):
+        """
+        Take step string, and add the appropriate R, V, O, M, or H step with appropriate parameters.
+        The step string input here is a single character (or character + number, for MTS)
+
+        Parameters
+        ----------
+        step_string : str
+            R, O, V, or Vn (where n is a nonnegative integer specifying force group)
+        measure_shadow_work : bool
+            Whether the steps should measure shadow work
+        measure_heat : bool
+            Whether the O step should measure heat
+        n_R : int
+            The number of R steps per integrator step
+        force_group_nV : dict
+            The number of V steps per integrator step per force group. {0: nV} if not mts
+        mts : bool
+            Whether the integrator is a multiple timestep integrator
+        """
+
+        if step_string == "O":
+            self.O_step(measure_heat)
+        elif step_string == "R":
+            self.R_step(measure_shadow_work, n_R)
+        elif step_string == "{":
+            self.addComputePerDof("xold", "x")
+            self.addComputePerDof("vold", "v")
+        elif step_string == "}":
```

Comment:
Should this include a check to see if `self._metropolize = True`, and if `{` has already appeared?